### PR TITLE
Optimize utils

### PR DESCRIPTION
**Pull request by jaceybronte**
Replacing keras with PyTorch for optimize_utils

### CODE REVIEW

File: 1.data-exploration/4.cell_line_summary_figure.ipynb [hunk 1]
```diff
@@ -42,7 +42,7 @@
    "outputs": [],
    "source": [
     "# Set i/o paths and files\n",
-    "data_dir <- file.path(\"../0.data-download/data\")\n",
+    "data_dir <- file.path(\"0.data-download/data\")\n",
```
Comment (gwaybio) [inline 1.data-exploration/4.cell_line_summary_figure.ipynb]: Why did this path change? I'm not sure if this will work now, if we are using relative paths

File: 2.train-VAE/optimize_utils.py [hunk 2]
```diff
@@ -1,96 +1,49 @@
...
-import random
+import torch
+import torch.nn as nn
+import torch.nn.functional as F
 
-random.seed(18)
 
 
-class HyperVAE(HyperModel):
+class HyperVAE(nn.Module):
```
Comment (gwaybio) [inline 2.train-VAE/optimize_utils.py]: Can you add a docstring? https://peps.python.org/pep-0257/

File: 2.train-VAE/optimize_utils.py [hunk 3]
```diff
@@ -1,96 +1,49 @@
...
-            ),
-            encoder_architecture=self.encoder_architecture,
-            decoder_architecture=self.decoder_architecture,
+        self, 
+        input_dim, 
+        latent_dim, 
+        beta):
+        super(HyperVAE, self).__init__()
+        self.encoder = nn.Sequential(
+            nn.Linear(input_dim, 128),
```
Comment (gwaybio) [inline 2.train-VAE/optimize_utils.py]: will we always want 128 or should this be a class argument?

File: 2.train-VAE/optimize_utils.py [hunk 4]
```diff
@@ -1,96 +1,49 @@
...
-                "encoder_batch_norm", default=self.encoder_batch_norm
-            ),
-            encoder_architecture=self.encoder_architecture,
-            decoder_architecture=self.decoder_architecture,
+        self, 
+        input_dim, 
+        latent_dim, 
+        beta):
+        super(HyperVAE, self).__init__()
+        self.encoder = nn.Sequential(
```
Comment (gwaybio) [inline 2.train-VAE/optimize_utils.py]: This appears to be a two layer encoder (and matched dimension with the decoder). Any reason for this? We typically only use one layer, but the evidence we cite for only using one is somewhat weak.

Comment (gwaybio) [reply 2.train-VAE/optimize_utils.py]: Is there also a way to make the number of layers a class parameter? (I'm not sure if this is even a helpful suggestion at this time, it might not be worth doing right now)

File: 2.train-VAE/optimize_utils.py [hunk 5]
```diff
@@ -1,96 +1,49 @@
...
+        )
+        self.latent_dim = latent_dim
+        self.beta = beta
 
-class CustomBayesianTunerCellPainting(BayesianOptimization):
-    # from https://github.com/keras-team/keras-tuner/issues/122#issuecomment-544648268
-    def run_trial(self, trial, *args, **kwargs):
-        kwargs["batch_size"] = trial.hyperparameters.Int("batch_size", 16, 128, step=32)
-        kwargs["epochs"] = trial.hyperparameters.Int("epochs", 5, 1000, step=100)
+    def reparameterize(self, mu, log_var):
```
Comment (gwaybio) [inline 2.train-VAE/optimize_utils.py]: These class methods should also have docstrings

File: 2.train-VAE/optimize_utils.py (hunk 2 above)
Comment (gwaybio) [reply 2.train-VAE/optimize_utils.py]: A docstring will also present an opportunity to cite references!

File: 2.train-VAE/optimize_utils.py [hunk 6]
```diff
@@ -1,96 +1,49 @@
...
-            encoder_architecture=self.encoder_architecture,
-            decoder_architecture=self.decoder_architecture,
+        self, 
+        input_dim, 
+        latent_dim, 
+        beta):
+        super(HyperVAE, self).__init__()
+        self.encoder = nn.Sequential(
+            nn.Linear(input_dim, 128),
+            nn.ReLU(),
```
Comment (gwaybio) [inline 2.train-VAE/optimize_utils.py]: IIRC, we typically add a batch normalization layer prior to the activation but after the linear

File: 2.train-VAE/optimize_utils.py [hunk 7]
```diff
@@ -1,96 +1,49 @@
 # THIS CODE WAS SOURCED FROM THE FOLLOWING URL: https://github.com/broadinstitute/cell-painting-vae/blob/master/scripts/optimize_utils.py
 
-from pyexpat import model
-from keras import backend as K
-from keras_tuner import HyperModel
-from keras_tuner.tuners import BayesianOptimization
-import sys
-from vae import VAE
-import random
+import torch
```
Comment (gwaybio) [inline 2.train-VAE/optimize_utils.py]: I think you can delete line 1 now (it departs from this URL now)

Also, as i'm looking through this function, it doesn't look like it belongs in the `optimize_utils.py` file (or at least it doesn't look like you'll be using it exclusively for optimization. It looks like it will work well for training and any other time you'd train/apply a VAE)

**Review by gwaybio — CHANGES_REQUESTED** [review]
I left several suggestions (mostly to do with documentation) and comments! We should go through another round of review prior to merging

**Comment by jaceybronte** [comment]
@gwaybio I figured out what was going on with Jupyter so reverted that line - also made a single layer for now

File: 1.data-exploration/scripts/nbconverted/3.gene_dependency_subsetting.py [hunk 8]
```diff
@@ -184,13 +184,14 @@
...
-variance_theshold = n.astype(float).min()
+variance_threshold = n.astype(float).min()
 
 # plotting variance density chart and marking the 1000 largest gene variation cutoff
 variance_density_plot = (
     p9.ggplot(variance_df, p9.aes(x="variance"))
     + p9.geom_density()
-    + p9.geom_vline(xintercept=variance_theshold, linetype="dashed", color="red")
+    + p9.geom_vline(xintercept=variance_threshold, linetype="dashed", color="red")
+    + p9.annotate(geom="text", x=variance_threshold + 0.13, y = 10, label="Top 1000 Gene Variances Threshold", color="red")
```
Comment (gwaybio) [inline 1.data-exploration/scripts/nbconverted/3.gene_dependency_subsetting.py:194]:
```suggestion
    + p9.annotate(geom="text", x=variance_threshold + 0.13, y=10, label="Top 1000 Gene Variances Threshold", color="red") 
```

Obviously a super nitpick, but formatting consistency will help in the long run :)

File: 2.train-VAE/betavae.py [hunk 9]
```diff
@@ -0,0 +1,115 @@
...
+        beta (float): Weight for the Kullback-Leibler divergence term in the loss function.
+    """
+    def __init__(
+        self, 
+        input_dim, 
+        #hidden_dim,
+        latent_dim, 
+        beta):
+        super(BetaVAE, self).__init__()
+        self.encoder = nn.Linear(input_dim, latent_dim * 2)
```
Comment (gwaybio) [inline 2.train-VAE/betavae.py]: why x2?

File: 2.train-VAE/betavae.py [hunk 10]
```diff
@@ -0,0 +1,115 @@
...
+    Args:
+        input_dim (int): Dimension of the input data.
+        hidden_dim (int): Dimension of the hidden layer. (this is utilized if two layers)
+        latent_dim (int): Dimension of the latent space.
+        beta (float): Weight for the Kullback-Leibler divergence term in the loss function.
+    """
+    def __init__(
+        self, 
+        input_dim, 
+        #hidden_dim,
```
Comment (gwaybio) [inline 2.train-VAE/betavae.py]: Looks like this is commented out? We typically avoid merging unfinished code when possible.

File: 2.train-VAE/betavae.py [hunk 11]
```diff
@@ -0,0 +1,115 @@
...
+        std = torch.exp(0.5 * log_var)
+        eps = torch.randn_like(std)
+        return mu + eps * std
+
+    def forward(self, x):
+        """
+        Forward pass through the VAE.
+
+        Args:
+
```
Comment (gwaybio) [inline 2.train-VAE/betavae.py]: Make sure docstrings are consistently formatted. We have to show you [precommit](https://pre-commit.com/) and autoformating! ([black](https://github.com/psf/black) is an example, but I'm not sure what the lab uses now)

File: 2.train-VAE/betavae.py [hunk 12]
```diff
@@ -0,0 +1,115 @@
...
+        Average loss over the test dataset.
+    """
+    model.eval()
+    test_loss = 0
+    with torch.no_grad():
+        for batch in test_loader:
+            data = batch[0]
+            recon, mu, log_var = model(data)
+            test_loss += model.loss_function(recon, data, mu, log_var).item()
+    return test_loss / len(test_loader.dataset)
```
Comment (gwaybio) [inline 2.train-VAE/betavae.py]: this nasty symbol means that you need to add a new line character to the end of your file :) (black/autoformatters will completely solve this)

File: 2.train-VAE/optimize_utils.py [hunk 13]
```diff
@@ -1,105 +1,11 @@
...
-
-
 def get_optimize_args():
     """
     Get arguments for the hyperparameter optimization procedure
     """
     import argparse
 
     parser = argparse.ArgumentParser()
+    parser.add_argument("-f","--fff", help="dummy argument", default="1")
```
Comment (gwaybio) [inline 2.train-VAE/optimize_utils.py:9]: Maybe vestigial code leftover from prototyping?

File: 2.train-VAE/betavae.py [hunk 14]
```diff
@@ -0,0 +1,115 @@
...
+            mu (Tensor): Mean of the latent Gaussian.
+            log_var (Tensor): Log variance of the latent Gaussian.
+
+        Returns:
+            Computed loss.
+        """
+        BCE = F.binary_cross_entropy(recon_x, x, reduction='sum')
+        KLD = -0.5 * torch.sum(1 + log_var - mu.pow(2) - log_var.exp())
+        return BCE + self.beta * KLD
+
```
Comment (gwaybio) [inline 2.train-VAE/betavae.py]: two spaces between function definitions :) (autoformating will solve this)

**Review by gwaybio — CHANGES_REQUESTED** [review]
Nearly there! Some procedural, but mostly formatting-related, comments.

I also left a book on your desk, which I hope provides some rationale for my nitpicky-ness and inspiration. Feel free to borrow and take it home with you if you'd like.

**Comment by jaceybronte** [comment]
@gwaybio I installed and hopefully successfully used Black - lmk if there is anything I need to do differently! The * 2 should be for outputting two separate vectors of size latent_dim that are then split with h.chunk(2, dim=-1) so that mu has the first and log_var has the second.

**Comment by gwaybio** [comment]
Formatting looks good!

> lmk if there is anything I need to do differently!

There appear to be other of my comments that remain unaddressed. Do you mean the PR is ready for re-review, or differently for formatting?

(FYI, typically, during a PR review, the author will respond to comments directly in each comment drop down text box. For instance, my question about `*2` would more appropriately fit in the reply functionality where I asked my question)

File: 2.train-VAE/optimize_utils.py (hunk 13 above)
Comment (jaceybronte) [reply 2.train-VAE/optimize_utils.py:9]: This resolves an ipykernel error with args - there may be another way to do it as well if that would be better!

**Comment by jaceybronte** [comment]
@gwaybio  I just added the batch normalization, sorry I missed that! I think it is ready for rereview now.

File: 2.train-VAE/optimize_utils.py (hunk 13 above)
Comment (gwaybio) [reply 2.train-VAE/optimize_utils.py:9]: oh interesting! This might be worth adding as a comment in the code to avoid confusion in the future

**Review by gwaybio — APPROVED** [review]
LGTM 🎉 feel free to merge at will
